File: site/public/assets/js/analytics.js
/* eSports Boost — first-party analytics beacon.
   -------------------------------------------------------------------------
   app.js already emits a clean funnel into window.dataLayer; until now nothing
   read it. This file is the other end: it mirrors those pushes to
   /api/collect, adds the events the dataLayer never had (page_view, configure,
   scroll, errors), and attaches the configurator state to every one of them —
   which is the whole point, because on this site the interesting question is
   never "how many visits" but "what were they configuring when they left".

   Anonymous by construction: a random id in localStorage, no cookie, no
   fingerprint, no personal field, and the server stores no IP. That keeps the
   whole pipeline inside the cookieless audience-measurement exemption, so
   there is no consent banner to add — please keep it that way.

   Opt-outs honoured: Global Privacy Control, and window.ESB_NO_ANALYTICS for
   local debugging. The /ops dashboard never reports on itself.

   Loaded after app.js, so window.esbState / esbQuote / dataLayer already exist.
*/
(function () {
  "use strict";

  var ENDPOINT = "/api/collect";
  var SESSION_IDLE = 30 * 60 * 1000;   // a new session after 30 min idle
  var FLUSH_MS = 2000;                 // batch window
  var MAX_QUEUE = 10;

  // The browser's own IANA timezone. No lookup, no IP, no third party — and
  // on a local or non-Vercel host it is the only thing that can tell you which
  // country a visitor is in.
  var TZ = (function () {
    try { return Intl.DateTimeFormat().resolvedOptions().timeZone || ""; }
    catch (e) { return ""; }
  })();

  var K_ANON = "esb.anon.v1";
  var K_SESS = "esb.sess.v1";
  var K_TOUCH = "esb.touch.v1";
  var K_INTERNAL = "esb.internal.v1";

  if (window.navigator && window.navigator.globalPrivacyControl) return;
  if (window.ESB_NO_ANALYTICS) return;
  if (location.pathname.indexOf("/ops") === 0) return;

  /* Internal traffic — our own browser, permanently. Test orders used to walk
     into the funnel as ordinary sessions, and over a few dozen clicks a handful
     of them moves the conversion rate further than anything a real visitor did:
     `paid` is simply "this session emitted purchase", so a test checkout counts
     exactly like a customer's. Load any page with ?esb_internal=1 to mark this
     browser and ?esb_internal=0 to clear it. A marked browser beacons NOTHING,
     so the store stays clean rather than carrying rows every future reader has
     to remember to filter out. It is per browser AND per profile — mark each
     one you test from, and re-mark after clearing site data.
     `window.ESB_NO_ANALYTICS` is still the one-page-load version of this. */
  function internalBrowser() {
    try {
      var q = new URLSearchParams(location.search);
      if (q.has("esb_internal")) {
        if (q.get("esb_internal") === "0") {
          localStorage.removeItem(K_INTERNAL);
          if (window.console) console.log("[esb] analytics ON — this browser is counted again");
        } else {
          localStorage.setItem(K_INTERNAL, "1");
          if (window.console) console.log("[esb] analytics OFF — this browser is marked internal");
        }
      }
      return localStorage.getItem(K_INTERNAL) === "1";
    } catch (e) { return false; }
  }
  if (internalBrowser()) return;

  /* ── storage helpers — analytics must never break a page ─────────────── */
  function get(key) {
    try { return JSON.parse(localStorage.getItem(key) || "null"); } catch (e) { return null; }
  }
  function put(key, val) {
    try { localStorage.setItem(key, JSON.stringify(val)); } catch (e) {}
  }
  function rid() {
    return (Math.random().toString(36).slice(2, 10) +
            Math.random().toString(36).slice(2, 6));
  }

  /* ── identity: one anonymous visitor id, one rolling session ─────────── */
  var anon = get(K_ANON);
  if (typeof anon !== "string" || !anon) { anon = rid(); put(K_ANON, anon); }

  var now = Date.now();
  var sess = get(K_SESS);
  var fresh = false;
  if (!sess || typeof sess.id !== "string" || (now - (sess.last || 0)) > SESSION_IDLE) {
    sess = { id: rid(), last: now, n: 0 };
    fresh = true;
  }

  /* ── first touch: where this visitor originally came from ────────────── */
  /* ⚠ Google Ads auto-tagging sends `gclid` and NOTHING else — never a utm_*
     — and the click routes through googleadservices.com, which lands here with
     an EMPTY referrer on mobile and in every in-app browser. Read through utm
     and referrer alone, a paid click therefore fell straight through to the
     literals below and was filed as `direct / none`: our own ad spend, in the
     one bucket that also holds real direct traffic, from which no cost per
     acquisition can be recovered. Hence the click ids.

     The id itself is READ AND DISCARDED — never stored, never sent. A click id
     is a unique per-click identifier the ad network can join back to a person,
     and keeping one would end the anonymous-by-construction promise at the top
     of this file, and with it the cookieless exemption that is why this site
     ships no consent banner. What is kept is the CHANNEL: `google / cpc`. */
  var PAID_CLICK = [
    ["gclid",   "google", "cpc"],   // Google Ads, ordinary web traffic
    ["gbraid",  "google", "cpc"],   // Google Ads, iOS web-to-app
    ["wbraid",  "google", "cpc"],   // Google Ads, iOS app-to-web
    ["msclkid", "bing",   "cpc"]    // Microsoft Ads
  ];

  function paidClick(q) {
    for (var p = 0; p < PAID_CLICK.length; p++) {
      if (q.get(PAID_CLICK[p][0])) {
        return { src: PAID_CLICK[p][1], med: PAID_CLICK[p][2] };
      }
    }
    return null;
  }

  var touch = get(K_TOUCH);
  (function () {
    var q = new URLSearchParams(location.search);
    var paid = paidClick(q);
    var tagged = !!(q.get("utm_source") || paid);

    /* Re-stamped on every TAGGED arrival, not only on a first visit. It used to
       be written once and kept for ever, so a browser that had ever seen the
       site stayed `direct` through every ad click it made afterwards — which
       drops exactly the returning visitors a remarketing campaign pays to bring
       back, and would have hidden the fix above from anyone already here. An
       UNTAGGED arrival still never overwrites a stored touch, so an ordinary
       organic return keeps the source it first came in on. */
    if (touch && !tagged) return;

    var ref = "";
    try {
      // Host only — never the full referring URL, which can carry a query.
      if (document.referrer) {
        var h = new URL(document.referrer).hostname;
        if (h && h !== location.hostname) ref = h.replace(/^www\./, "");
      }
    } catch (e) {}

    touch = {
      src: q.get("utm_source") || (paid && paid.src) || ref || "direct",
      med: q.get("utm_medium") || (paid && paid.med) || (ref ? "referral" : "none"),
      cmp: q.get("utm_campaign") || "",
      ref: ref
    };
    put(K_TOUCH, touch);
  })();

  /* ── the configurator snapshot carried by every event ────────────────── */
  function snapshot() {
    if (!window.esbState || !window.esbQuote) return null;
    var s, q;
    try { s = window.esbState(); q = window.esbQuote(s); } catch (e) { return null; }
    if (!s) return null;
    var cfg = {
      game: s.game, service: s.service, from: s.from, to: s.to,
      mode: s.mode, region: s.region, addons: (s.addons || []).slice(0, 10)
    };
    if (s.promo) cfg.promo = s.promo;
    if (s.service === "wins") cfg.wins = s.wins;
    if (s.service === "placements") cfg.placements = s.placements;
    if (q) {
      cfg.total = q.total;
      cfg.summary = q.summary || "";
      if (q.invalid) cfg.invalid = true;
    }
    return cfg;
  }

  /* ── queue & flush ───────────────────────────────────────────────────── */
  var queue = [];
  var timer = null;

  function flush() {
    if (timer) { clearTimeout(timer); timer = null; }
    if (!queue.length) return;
    var body = JSON.stringify({ events: queue.splice(0, queue.length) });
    try {
      if (navigator.sendBeacon) {
        // A Blob with an explicit type keeps this a simple request (no preflight)
        // and lets it survive the pagehide that usually kills the last event.
        navigator.sendBeacon(ENDPOINT, new Blob([body], { type: "application/json" }));
        return;
      }
    } catch (e) {}
    try {
      fetch(ENDPOINT, {
        method: "POST", body: body, keepalive: true,
        headers: { "Content-Type": "application/json" }
      }).catch(function () {});
    } catch (e) {}
  }

  function emit(name, extra) {
    sess.n += 1;
    sess.last = Date.now();
    put(K_SESS, sess);

    var ev = {
      e: name, a: anon, s: sess.id, n: sess.n,
      p: location.pathname,
      src: touch.src, med: touch.med, cmp: touch.cmp, ref: touch.ref,
      lang: (navigator.language || "").slice(0, 12),
      tz: TZ
    };
    var cfg = snapshot();
    if (cfg) ev.cfg = cfg;
    if (extra) {
      if (extra.val != null) ev.val = extra.val;
      if (extra.meta) ev.meta = extra.meta;
      if (extra.cfg) ev.cfg = extra.cfg;
    }
    queue.push(ev);

    if (queue.length >= MAX_QUEUE) flush();
    else if (!timer) timer = setTimeout(flush, FLUSH_MS);
  }
  window.esbEmit = emit;

  /* ── bridge: every dataLayer push app.js already makes ───────────────── */
  // select_item / add_to_cart are configuration changes under other names —
  // the signature watcher below is the single, de-duplicated source for those,
  // so mirroring them here too would double-count every re-quote.
  // The account flow rides the same bridge as the funnel: app.js pushes it to
  // dataLayer at each step (see initHeader) and it arrives here by name. What
  // crosses is the STEP and its outcome — mode, method, reason — never the email
  // or the name the visitor typed. Those belong to the accounts store, which is
  // a separate store precisely so this one stays anonymous.
  var BRIDGE = {
    view_item: "view_item", select_promotion: "select_promotion",
    view_promotion: "view_promotion",
    begin_checkout: "begin_checkout", add_payment_info: "add_payment_info",
    purchase: "purchase", generate_lead: "generate_lead",
    auth_open: "auth_open", oauth_start: "oauth_start", sign_up: "sign_up",
    login: "login", logout: "logout", auth_error: "auth_error",
    /* The accounts shop's own three steps. They are house names, deliberately
       NOT `select_item` / `view_item_list`: `select_item` is pushed by the
       configurator on every re-quote and is kept out of this bridge on purpose
       (see the note on `configure` below), so borrowing it here would start
       double-counting boosting re-quotes as account picks. */
    account_shop: "account_shop", account_server: "account_server",
    account_tiers: "account_tiers"
  };

  // Only these keys are carried, and each is stringified — a payload key that
  // is not named here never reaches the store, which is what stops a future
  // caller passing `email` into track() and having it silently persisted.
  // `shard`, `tiers` and `stock` are the accounts shop's: a region code and two
  // counts. Product facts about the page, never about the person — which is the
  // only test a key added here has to pass.
  var META_KEYS = ["transaction_id", "method", "mode", "reason", "promotion",
                   "shard", "tiers", "stock"];

  function bridge(payload) {
    if (!payload || typeof payload !== "object") return;
    var name = BRIDGE[payload.event];
    if (!name) return;
    var extra = {};
    if (typeof payload.value === "number") extra.val = payload.value;
    var meta = {};
    for (var k = 0; k < META_KEYS.length; k++) {
      var key = META_KEYS[k];
      /* `!= null` and not a truthiness test: the accounts shop's `stock` and
         `tiers` are counts, and ZERO is the reading worth having — a shard
         whose whole board is sold out is the one case this beacon exists to
         catch. An empty string is still dropped. */
      if (payload[key] != null && payload[key] !== "") meta[key] = String(payload[key]);
    }
    if (Object.keys(meta).length) extra.meta = meta;
    emit(name, extra);
  }

  var dl = window.dataLayer = window.dataLayer || [];
  for (var i = 0; i < dl.length; i++) bridge(dl[i]);   // anything already pushed
  var push = dl.push.bind(dl);
  dl.push = function (payload) {
    var out = push(payload);
    try { bridge(payload); } catch (e) {}
    return out;
  };

  /* ── configure: fires only when the quote actually changed ───────────── */
  // Watching the DOM contract (data-sel / data-service / data-mode / …) rather
  // than app.js internals, then comparing signatures, so a click that lands on
  // the same configuration is not a re-quote.
  var lastSig = JSON.stringify(snapshot());
  var configTimer = null;

  function maybeConfigure() {
    if (configTimer) clearTimeout(configTimer);
    configTimer = setTimeout(function () {
      var sig = JSON.stringify(snapshot());
      if (sig === lastSig || sig === "null") return;
      lastSig = sig;
      emit("configure");
    }, 350);
  }

  /* ── engage: the first real touch of a configurator control ────────── */
  /* The funnel's second step used to be `view_item`, which app.js fires at the
     end of init() on any page carrying [data-configurator] — the homepage
     Best Sellers band included. It needs NO interaction, so every single
     homepage load reported "opened a configurator", carrying the page's own
     default climb as the session value. The step measured whether the HTML
     rendered, and a bounce was indistinguishable from a buyer.

     This is the honest version of that claim: the visitor touched a control.
     Deliberately NOT the same thing as `configure`, which needs the quote to
     actually change — tapping the tab you are already on is engagement, not a
     re-quote. Once per session; `view_item` still fires for GA4 and Ads. */
  var engaged = false;
  function maybeEngage() {
    if (engaged) return;
    engaged = true;
    emit("engage");
  }

  var WATCH = "[data-sel],[data-service],[data-mode],[data-addon],[data-stepper]," +
              "[data-ladder],[data-promo-apply],[data-panel]";
  document.addEventListener("change", function (e) {
    if (e.target && e.target.closest && e.target.closest(WATCH)) {
      maybeEngage(); maybeConfigure();
    }
  }, true);
  document.addEventListener("click", function (e) {
    if (e.target && e.target.closest && e.target.closest(WATCH)) {
      maybeEngage(); maybeConfigure();
    }
  }, true);

  /* ── scroll depth ────────────────────────────────────────────────────── */
  var marks = [25, 50, 75, 100], hit = {}, scrollTimer = null;
  window.addEventListener("scroll", function () {
    if (scrollTimer) return;
    scrollTimer = setTimeout(function () {
      scrollTimer = null;
      var h = document.documentElement.scrollHeight - window.innerHeight;
      if (h <= 0) return;
      var pct = Math.min(100, Math.round((window.scrollY / h) * 100));
      for (var j = 0; j < marks.length; j++) {
        if (pct >= marks[j] && !hit[marks[j]]) {
          hit[marks[j]] = 1;
          emit("scroll", { meta: { pct: marks[j] } });
        }
      }
    }, 500);
  }, { passive: true });

  /* ── errors ──────────────────────────────────────────────────────────── */
  window.addEventListener("error", function (e) {
    emit("js_error", {
      meta: {
        message: String((e && e.message) || "error").slice(0, 160),
        source: String((e && e.filename) || "").slice(0, 120),
        line: (e && e.lineno) || 0
      }
    });
  });

  /* ── lifecycle ───────────────────────────────────────────────────────── */
  // Whether the visitor arrived already signed in. Without it a timeline that
  // shows no login is ambiguous between a guest and somebody who logged in
  // last week and never signed out — and reading the second as the first
  // undercounts the account flow every time. It is a boolean about the browser,
  // not about the person: no id, no email, nothing that survives a sign-out.
  function signedIn() {
    try { return localStorage.getItem("esb.session.v1") ? "in" : "out"; }
    catch (e) { return ""; }
  }

  if (fresh) emit("session_start", { meta: { account: signedIn() } });
  emit("page_view");

  /* ── leaving — the one event that makes a duration real ──────────── */
  /* A session's duration is the gap between its first and its last event, and
     every event above fires within milliseconds of load. So a visitor who read
     a landing page for two minutes and left without scrolling a quarter of the
     way down recorded 0s — the same figure as an instant bounce, which is the
     single distinction paid traffic has to be judged on. Nothing else can
     supply it: there is no heartbeat and `scroll` only fires at 25% depth.

     Once per page load, on whichever signal comes first: `pagehide` does not
     fire at all on some mobile browsers, and `hidden` is what a phone actually
     sends when the visitor switches away. Firing it early costs nothing — the
     duration reads the LAST event, so anything done after coming back extends
     it again. */
  var left = false;
  function leaving() {
    if (!left) {
      left = true;
      emit("page_exit", { meta: { sec: Math.round((Date.now() - now) / 1000) } });
    }
    flush();
  }

  document.addEventListener("visibilitychange", function () {
    if (document.visibilityState === "hidden") leaving();
  });
  window.addEventListener("pagehide", leaving);
})();

File: site/src/analytics.py
# -*- coding: utf-8 -*-
"""First-party analytics: the event store and the ingest endpoint.

The sales site already emits a clean funnel (`view_item` → `begin_checkout` →
`purchase`, see public/assets/js/app.js) into `window.dataLayer`, but nothing
ever read that array. This module is the other end of the pipe: it takes those
events from `/api/collect` and puts them somewhere they survive a request.

Two stores, chosen by environment — same interface, no third-party packages:

  * **Upstash Redis** over its REST API (stdlib `urllib`) when
    `UPSTASH_REDIS_REST_URL` + `UPSTASH_REDIS_REST_TOKEN` are set. This is the
    production path: Vercel's filesystem is ephemeral, so a file would lose
    every event the moment the function froze.
  * **A local NDJSON file** otherwise, so `serve.py` collects for real during
    development with nothing to configure.

Privacy: the schema is anonymous by construction. No IP is stored, no cookie is
set, no field carries a name or an email — the visitor id is a random string the
browser mints for itself. Country comes from the edge header Vercel already
attaches, never from an IP lookup we perform. That keeps this out of
consent-banner territory, which is also why it is worth defending: do not add a
field here that could identify a person.

`/api/collect` is a public, unauthenticated endpoint — anything reaching it is
hostile until proven otherwise. Everything is allowlisted, length-capped and
type-checked below; nothing is stored that did not survive `_clean_event()`.
"""
import json
import os
import re
import time
import urllib.error
import urllib.parse
import urllib.request

import geo

# ── limits ────────────────────────────────────────────────────────────────
MAX_BODY = 64 * 1024      # a beacon payload is tiny; cap it hard
MAX_BATCH = 20            # events per POST
MAX_EVENTS = int(os.environ.get("ANALYTICS_MAX_EVENTS", "50000") or 50000)
HTTP_TIMEOUT = 5

LIST_KEY = "esb:ev"

# Every event the client is allowed to name. An unknown name is dropped rather
# than stored — otherwise the store is a free-form public write target.
ALLOWED_EVENTS = {
    "session_start", "page_view", "view_item", "configure", "select_promotion",
    # The mystery-discount modal was shown. The mystery store only records cards
    # people actually opened, so without this there is no denominator for the
    # flow — "N leads" reads as a win with no idea what it cost in interruptions.
    "view_promotion",
    "add_to_cart", "begin_checkout", "add_payment_info", "purchase",
    "generate_lead", "checkout_error", "js_error", "scroll", "engage",
    # Emitted once per page as the visitor leaves — pagehide, or the tab going
    # hidden, whichever fires first. It is what gives a session a real duration:
    # every other event on a landing page fires within milliseconds of load, so
    # without it a two-minute read and an instant bounce both record 0s, which
    # is the one distinction paid traffic has to be judged on. `meta.sec` is the
    # dwell the page measured for itself.
    "page_exit",
    # The account flow, end to end: the panel opening, a departure for an OAuth
    # provider, the two outcomes, the sign-out, and every refusal in between.
    # Without these a session that creates an account looks exactly like one
    # that browsed past the header — the sign-up shows up in the /ops Accounts
    # list with no way back to what the visitor was doing when they made it.
    # These record THAT an account step happened, never WHOSE: the email and the
    # name live in the accounts store, which is a separate store for precisely
    # that reason. Nothing below may ever carry either — see the note at the top
    # of this file, and `meta` is capped and stringified like every other field.
    "auth_open", "oauth_start", "sign_up", "login", "logout", "auth_error",
    # The accounts shop (`/accounts`), which is NOT a configurator and so fires
    # none of the funnel's own events. Between `page_view` and `begin_checkout`
    # that page emitted nothing at all, which made three very different visits
    # indistinguishable in /ops: somebody who read the hero and left, somebody
    # who picked a server and balked at the board, and somebody whose browser
    # loaded the HTML but never mounted the shop. In order:
    #   account_shop    the shop mounted and is interactive (the denominator,
    #                   and the one thing `page_view` cannot say — a page_view
    #                   on /accounts with no account_shop behind it is a broken
    #                   render, not a bounce)
    #   account_server  step 1 answered: a shard was picked
    #   account_tiers   step 2 on screen with real cards, once per shard
    # `meta` carries the shard code and two counts (`tiers`, `stock`) — product
    # facts about the page, never about the person, like everything else here.
    "account_shop", "account_server", "account_tiers",
}

# Configurator fields mirrored from the order state in app.js. Anything else in
# a posted `cfg` is discarded.
CFG_STR = ("game", "service", "from", "to", "mode", "region", "promo", "summary")
CFG_NUM = ("total", "wins", "placements", "steps")

_ID_RE = re.compile(r"^[A-Za-z0-9_.-]{1,40}$")
_CTRL_RE = re.compile(r"[\x00-\x1f\x7f]")


# ══════════════════════════════════════════════════════════════════════════
#  stores
# ══════════════════════════════════════════════════════════════════════════
def upstash_config():
    """(url, token) when Upstash is configured, else (None, None)."""
    url = os.environ.get("UPSTASH_REDIS_REST_URL", "").strip().rstrip("/")
    token = os.environ.get("UPSTASH_REDIS_REST_TOKEN", "").strip()
    return (url, token) if url and token else (None, None)


def store_name():
    return "upstash" if upstash_config()[0] else "file"


def log_path():
    return os.environ.get("ANALYTICS_LOG", "").strip() or os.path.join(
        os.path.dirname(os.path.abspath(__file__)), "..", "analytics.ndjson")


class StoreError(Exception):
    pass


def _upstash(commands):
    """Run a pipeline of Redis commands against Upstash's REST API.

    `commands` is a list of arg-lists, e.g. [["LPUSH", key, val], ["LTRIM", …]].
    Returns the list of results. Raises StoreError on any transport failure —
    callers decide whether that is fatal (reading) or swallowed (writing).
    """
    url, token = upstash_config()
    if not url:
        raise StoreError("Upstash not configured")
    body = json.dumps(commands).encode()
    req = urllib.request.Request(
        url + "/pipeline", data=body, method="POST",
        headers={"Authorization": "Bearer " + token,
                 "Content-Type": "application/json"})
    try:
        with urllib.request.urlopen(req, timeout=HTTP_TIMEOUT) as r:
            out = json.loads(r.read().decode() or "[]")
    except urllib.error.HTTPError as e:
        raise StoreError("Upstash HTTP %s: %s" % (e.code, e.read()[:200]))
    except Exception as e:                                    # noqa: BLE001
        raise StoreError("Upstash unreachable: %s" % e)
    for item in out if isinstance(out, list) else []:
        if isinstance(item, dict) and item.get("error"):
            raise StoreError(str(item["error"])[:200])
    return [i.get("result") if isinstance(i, dict) else i
            for i in (out if isinstance(out, list) else [])]


def append(events):
    """Persist a list of cleaned events. Never raises: a dropped analytics
    event must never turn into a 500 on a page the customer is looking at."""
    if not events:
        return 0
    lines = [json.dumps(e, separators=(",", ":")) for e in events]
    if upstash_config()[0]:
        try:
            # Newest-first list, trimmed to the cap so the store is bounded.
            _upstash([["LPUSH", LIST_KEY] + lines,
                      ["LTRIM", LIST_KEY, 0, MAX_EVENTS - 1]])
            return len(lines)
        except StoreError:
            return 0
    try:
        with open(log_path(), "a") as f:
            f.write("\n".join(lines) + "\n")
        return len(lines)
    except OSError:
        return 0


def read(limit=MAX_EVENTS):
    """Return up to `limit` stored events, newest first."""
    limit = max(1, min(int(limit or MAX_EVENTS), MAX_EVENTS))
    if upstash_config()[0]:
        try:
            res = _upstash([["LRANGE", LIST_KEY, 0, limit - 1]])
        except StoreError:
            return []
        rows = res[0] if res else []
    else:
        try:
            with open(log_path()) as f:
                rows = f.read().splitlines()
        except OSError:
            return []
        rows = rows[-limit:][::-1]          # file is oldest-first; flip it
    out = []
    for row in rows or []:
        try:
            ev = json.loads(row)
        except (ValueError, TypeError):
            continue
        if isinstance(ev, dict):
            out.append(ev)
    return out


def clear():
    """Wipe the store. Used by the seeder; never exposed over HTTP."""
    if upstash_config()[0]:
        try:
            _upstash([["DEL", LIST_KEY]])
        except StoreError:
            return False
        return True
    try:
        os.remove(log_path())
    except OSError:
        pass
    return True


def count():
    if upstash_config()[0]:
        try:
            res = _upstash([["LLEN", LIST_KEY]])
        except StoreError:
            return 0
        return int(res[0] or 0) if res else 0
    try:
        with open(log_path()) as f:
            return sum(1 for _ in f)
    except OSError:
        return 0


# ══════════════════════════════════════════════════════════════════════════
#  ingest
# ══════════════════════════════════════════════════════════════════════════
def _s(v, n=120):
    """Any client value → a bounded, control-char-free string."""
    if v is None or isinstance(v, (dict, list)):
        return ""
    return _CTRL_RE.sub("", str(v))[:n].strip()


def _num(v, lo=-1e9, hi=1e9):
    try:
        f = float(v)
    except (TypeError, ValueError):
        return None
    if f != f or f in (float("inf"), float("-inf")):   # NaN / ±inf
        return None
    return max(lo, min(hi, round(f, 2)))


def _id(v):
    v = _s(v, 40)
    return v if _ID_RE.match(v) else ""


def device_class(ua):
    ua = (ua or "").lower()
    if "ipad" in ua or ("android" in ua and "mobile" not in ua) or "tablet" in ua:
        return "tablet"
    if "mobi" in ua or "iphone" in ua or "android" in ua:
        return "mobile"
    return "desktop"


def _clean_cfg(cfg):
    if not isinstance(cfg, dict):
        return None
    out = {}
    for k in CFG_STR:
        v = _s(cfg.get(k), 80)
        if v:
            out[k] = v
    for k in CFG_NUM:
        v = _num(cfg.get(k), 0, 100000)
        if v is not None:
            out[k] = v
    addons = cfg.get("addons")
    if isinstance(addons, list):
        out["addons"] = [_s(a, 40) for a in addons[:10] if _s(a, 40)]
    if cfg.get("invalid"):
        out["invalid"] = True
    return out or None


def _clean_event(raw, now, ctx):
    """One posted event → the stored record, or None if it fails the gate."""
    if not isinstance(raw, dict):
        return None
    name = _s(raw.get("e") or raw.get("event"), 40)
    if name not in ALLOWED_EVENTS:
        return None
    anon, sess = _id(raw.get("a")), _id(raw.get("s"))
    if not anon or not sess:
        return None

    ev = {"t": now, "e": name, "a": anon, "s": sess}

    seq = _num(raw.get("n"), 0, 10000)
    if seq is not None:
        ev["n"] = int(seq)
    for key, src, n in (("p", "p", 160), ("ref", "ref", 80), ("src", "src", 60),
                        ("med", "med", 40), ("cmp", "cmp", 60), ("lang", "lang", 12),
                        ("tz", "tz", 48)):
        v = _s(raw.get(src), n)
        if v:
            ev[key] = v
    val = _num(raw.get("val"), 0, 100000)
    if val is not None:
        ev["val"] = val
    cfg = _clean_cfg(raw.get("cfg"))
    if cfg:
        ev["cfg"] = cfg
    meta = raw.get("meta")
    if isinstance(meta, dict):
        m = {}
        for k in list(meta)[:8]:
            key = _s(k, 24)
            v = meta[k]
            if key:
                m[key] = _num(v, -1e9, 1e9) if isinstance(v, (int, float)) else _s(v, 160)
        if m:
            ev["meta"] = m

    # Server-side context — the client cannot set or spoof these.
    ev["dev"] = ctx["dev"]
    # Country is resolved here rather than trusted from the body: the edge
    # header wins, and the browser's timezone/locale only fill in behind it.
    co = geo.country(ctx["co"], ev.get("tz"), ev.get("lang"))
    if co:
        ev["co"] = co
        ev["cosrc"] = geo.source(ctx["co"], ev.get("tz"), ev.get("lang"))
    return ev


def process_collect(raw, header_get):
    """POST /api/collect → (status, payload).

    Body is `{"events": [...]}` or a bare list. Returns 204 with an empty body
    on success — the browser's `sendBeacon` ignores the response either way, and
    a 204 keeps the endpoint useless as an oracle.
    """
    if not raw or len(raw) > MAX_BODY:
        return 204, None
    try:
        body = json.loads(raw.decode("utf-8", "replace") or "{}")
    except ValueError:
        return 204, None
    events = body.get("events") if isinstance(body, dict) else body
    if not isinstance(events, list):
        return 204, None

    get = header_get or (lambda *_a, **_k: "")
    country = _s(get("x-vercel-ip-country") or "", 2).upper()
    ctx = {
        "dev": device_class(get("user-agent") or ""),
        "co": country if len(country) == 2 and country.isalpha() else "",
    }
    now = int(time.time())
    clean = [c for c in (_clean_event(e, now, ctx) for e in events[:MAX_BATCH]) if c]
    append(clean)
    return 204, None
